rfd/0163-vnet.md

```diff
@@ -0,0 +1,996 @@
+
+Some users may want to use custom domain names for their apps.
+Maybe they currently access an app on their VPN at `api.example.com` and
+they want to migrate off their VPN to Teleport without changing all their
+software and scripts that currently reference that name.
+To do this they can set the `spec.public_addr` field in the app resource to
+`api.example.com`.
+
+We do not want the VNet DNS server to be resolving all DNS requests on the host
+and searching all clusters for apps with a matching `public_addr` on every
+request, this would introduce too much latency to unrelated DNS queries.
+To avoid this we will add cluster-level configuration of custom DNS zones which
+will be used to narrow the scope of:
+
+1. The split DNS configuration on the host that determines which DNS queries
+   will be sent to VNet.
+2. Which Teleport clusters that should be searched for matching apps when a DNS
+   query is received.
+
+When custom addresses are desired the Teleport admin can configure a custom DNS
+zone with a new `vnet` Teleport configuration resource:
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  custom_dns_zones:
+  - suffix: example.com
+  - suffix: goteleport.com
+```
+
+In this example, whenever the user is logged in to the cluster where this `vnet`
+configuration resource is present, VNet will install itself as the DNS
+nameserver for subdomains of `example.com` and `goteleport.com`.
+When the DNS server receives a query for any subdomain of those suffixes, it
+will search the cluster for an app with that `public_addr`.
+
+#### Leaf and parallel clusters
+
+A user may be logged in to multiple Teleport Clusters at once, those may be leaf
+clusters or completely separate clusters.
+
+When resolving a DNS query, VNet will search all clusters where the proxy
+address or any custom DNS zone in the cluster is a suffix of the queried name.
+
+For apps in leaf clusters, VNet will still use the `public_addr` of the app even
+if it is the default `public_addr` which includes the address of the leaf
+cluster.
+This differs conceptually from non-VNet app access where leaf apps would use the
+name `<app-name>.<root-proxy-address>`.
+This is convenient for the implementation and helps to avoid conflicts in cases
+where an app with the same name exists in both a root and leaf cluster.
+If both apps used `<app-name>.<root-proxy-address>` there would be no way to
+resolve which to connect to like there is in the web where a TLS client
+certificate or a session cookie can be used to differentiate the two.
+
+To avoid conflicts between multiple clusters where the `spec.public_addr` of two
+apps are identical, VNet will refuse to resolve names for any overlapping custom
+DNS zones across clusters the user is currently logged in to.
+
+In the corner case where a user is logged in to the same cluster both as a leaf
+and a root, VNet will prefer to proxy to apps in that cluster as a root.
+
+### DNS queries and IP address assignment
```

Comment:
I don't see any mention on what audit events will be captured.  It seems like IP Assignment will be a critical one for possible review into activity.  Can you add a section on audit logging and monitoring?

Reply:
Thank you, I will move my comment there